File: src/api/domains/rmrb.ts
import * as d from '../_shared'

export const apiRmrb = {
  // ===== 人民日报 / 申论 =====

  getRmrbMeta(): Promise<d.ApiRes<d.ShenlunMeta>> {
    return d.isMock ? d.mockService.getRmrbMeta() : d.request('/api/rmrb/meta')
  },

  createRmrbSkeletonTemplate(data: {
    name: string
    description?: string
    mode?: string
    structure?: d.ShenlunSkeletonStructure
    sortOrder?: number
    isEnabled?: boolean
  }): Promise<d.ApiRes<d.ShenlunSkeletonTemplate>> {
    return d.isMock
      ? d.mockService.createRmrbSkeletonTemplate(data)
      : d.request('/api/rmrb/skeleton-templates', { method: 'POST', data })
  },

  createRmrbTermCategory(data: {
    name: string
    kind?: 'term' | 'verb' | string
    sortOrder?: number
    isEnabled?: boolean
  }): Promise<d.ApiRes<d.ShenlunTermCategory>> {
    return d.isMock
      ? d.mockService.createRmrbTermCategory(data)
      : d.request('/api/rmrb/term-categories', { method: 'POST', data })
  },

  getRmrbStats(): Promise<d.ApiRes<d.ShenlunStats>> {
    return d.isMock ? d.mockService.getRmrbStats() : d.request('/api/rmrb/stats')
  },

  listRmrbArticles(tag?: string): Promise<d.ApiRes<d.RmrbArticle[]>> {
    if (d.isMock) return d.mockService.listRmrbArticles(tag)
    const q = tag ? `?tag=${encodeURIComponent(tag)}` : ''
    return d.request(`/api/rmrb/articles${q}`)
  },

  getRmrbArticle(id: string): Promise<d.ApiRes<d.RmrbArticle>> {
    return d.isMock ? d.mockService.getRmrbArticle(id) : d.request(`/api/rmrb/articles/${id}`)
  },

  listRmrbMines(): Promise<d.ApiRes<d.ShenlunMineLog[]>> {
    return d.isMock ? d.mockService.listRmrbMines() : d.request('/api/rmrb/mines')
  },

  getRmrbMine(id: string): Promise<d.ApiRes<d.ShenlunMineLog>> {
    return d.isMock ? d.mockService.getRmrbMine(id) : d.request(`/api/rmrb/mines/${id}`)
  },

  getRmrbMineByDate(date: string): Promise<d.ApiRes<d.ShenlunMineLog>> {
    return d.isMock
      ? d.mockService.getRmrbMineByDate(date)
      : d.request(`/api/rmrb/mines/by-date/${date}`)
  },

  upsertRmrbMine(data: {
    mineDate?: string
    articleId?: string | null
    articleTitle?: string
    sourceExcerpt?: string
    argumentChain?: string
    templateSentence?: string
    terms?: Array<d.ShenlunMineTermItem | string>
    quotes?: d.ShenlunQuoteItem[]
    verbs?: d.ShenlunVerbItem[]
    argument?: d.ShenlunArgumentSkeleton
    templates?: d.ShenlunTemplateItem[]
  }): Promise<d.ApiRes<d.ShenlunMineLog>> {
    return d.isMock
      ? d.mockService.upsertRmrbMine(data)
      : d.request('/api/rmrb/mines', { method: 'POST', data })
  },

  updateRmrbMine(
    id: string,
    data: Partial<{
      articleId: string | null
      articleTitle: string
      sourceExcerpt: string
      argumentChain: string
      templateSentence: string
      terms: Array<d.ShenlunMineTermItem | string>
      quotes: d.ShenlunQuoteItem[]
      verbs: d.ShenlunVerbItem[]
      argument: d.ShenlunArgumentSkeleton
      templates: d.ShenlunTemplateItem[]
    }>,
  ): Promise<d.ApiRes<d.ShenlunMineLog>> {
    return d.isMock
      ? d.mockService.updateRmrbMine(id, data)
      : d.request(`/api/rmrb/mines/${id}`, { method: 'PUT', data })
  },

  deleteRmrbMine(id: string): Promise<d.ApiRes<{ ok: boolean }>> {
    return d.isMock
      ? d.mockService.deleteRmrbMine(id)
      : d.request(`/api/rmrb/mines/${id}`, { method: 'DELETE' })
  },

  listRmrbTerms(
    status?: 'learning' | 'mastered',
    category?: string,
  ): Promise<d.ApiRes<d.ShenlunNormTerm[]>> {
    if (d.isMock) return d.mockService.listRmrbTerms()
    const qs = new URLSearchParams()
    if (status) qs.set('status', status)
    if (category) qs.set('category', category)
    const q = qs.toString() ? `?${qs}` : ''
    return d.request(`/api/rmrb/terms${q}`)
  },

  addRmrbTerm(data: {
    term: string
    category?: string
    usageNote?: string
    sourceTitle?: string
    exampleSentence?: string
    articleId?: string | null
  }): Promise<d.ApiRes<d.ShenlunNormTerm>> {
    return d.isMock
      ? d.mockService.addRmrbTerm(data)
      : d.request('/api/rmrb/terms', { method: 'POST', data })
  },

  updateRmrbTerm(
    id: string,
    data: Partial<{
      category: string
      usageNote: string
      exampleSentence: string
      familiarity: number
      mastered: boolean
      sourceTitle: string
    }>,
  ): Promise<d.ApiRes<d.ShenlunNormTerm>> {
    return d.isMock
      ? d.mockService.updateRmrbTerm(id, data)
      : d.request(`/api/rmrb/terms/${id}`, { method: 'PUT', data })
  },

  deleteRmrbTerm(id: string): Promise<d.ApiRes<{ ok: boolean }>> {
    return d.isMock
      ? d.mockService.deleteRmrbTerm(id)
      : d.request(`/api/rmrb/terms/${id}`, { method: 'DELETE' })
  },

  listRmrbDrills(
    drillType?: 'sentence' | 'imitate' | 'oral',
  ): Promise<d.ApiRes<d.ShenlunDrillLog[]>> {
    if (d.isMock) return d.mockService.listRmrbDrills()
    const q = drillType ? `?drill_type=${drillType}` : ''
    return d.request(`/api/rmrb/drills${q}`)
  },

  addRmrbDrill(data: {
    drillType: 'sentence' | 'imitate' | 'oral'
    content: string
    prompt?: string
    refMineId?: string | null
    refTermIds?: string[]
  }): Promise<d.ApiRes<d.ShenlunDrillLog>> {
    return d.isMock
      ? d.mockService.addRmrbDrill(data)
      : d.request('/api/rmrb/drills', { method: 'POST', data })
  },
}

File: server/app/schemas/rmrb.py
"""Pydantic schema · 域模块

按业务域拆分，统一由 app/schemas/__init__.py re-export，保持 from app.schemas import X 兼容。
"""
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field
class RmrbArticleOut(BaseModel):
    id: str
    title: str
    source: str
    sourceUrl: str = ""
    publishDate: str
    summary: str
    content: str
    tags: list[str] = []
    isPublished: bool
    sortOrder: int
    readCount: int
    createdAt: datetime
    updatedAt: datetime


class RmrbArticleCreate(BaseModel):
    title: str
    source: str = "人民时评"
    sourceUrl: str = ""
    publishDate: str = ""
    summary: str = ""
    content: str = ""
    tags: list[str] = []
    isPublished: bool = True
    sortOrder: int = 0


class RmrbArticleUpdate(BaseModel):
    title: str | None = None
    source: str | None = None
    sourceUrl: str | None = None
    publishDate: str | None = None
    summary: str | None = None
    content: str | None = None
    tags: list[str] | None = None
    isPublished: bool | None = None
    sortOrder: int | None = None


class ShenlunMineTermItem(BaseModel):
    term: str
    category: str = "其他"
    plainWord: str = ""


class ShenlunQuoteItem(BaseModel):
    text: str = ""
    source: str = ""  # 来源，如：清代万斯大
    meaning: str = ""  # 释义


class ShenlunVerbItem(BaseModel):
    verb: str = ""
    usage: str = ""
    category: str = "其他"


class ShenlunArgumentFieldValue(BaseModel):
    key: str
    label: str = ""
    content: str = ""


class ShenlunArgumentPoint(BaseModel):
    title: str = ""  # 分论点正文
    claim: str = ""  # 兼容旧数据
    evidence: str = ""
    summary: str = ""
    method: str = ""  # 论证方法名，如：点例排比 + 类比延伸
    methodNote: str = ""  # 方法说明
    template: str = ""  # 套用模板


class ShenlunArgumentSkeleton(BaseModel):
    templateId: str = ""
    templateName: str = ""
    mode: str = "points"  # linear | points
    overview: str = ""  # 总论点
    conclusion: str = ""  # 总结
    overviewMethod: str = ""  # 总论点论证方法
    overviewTemplate: str = ""  # 总论点论证模板
    fields: list[ShenlunArgumentFieldValue] = []
    points: list[ShenlunArgumentPoint] = []


class ShenlunTemplateItem(BaseModel):
    type: str = "dialectic"  # sentence type code
    typeName: str = ""
    original: str = ""
    template: str = ""
    imitate: str = ""


class ShenlunMineLogOut(BaseModel):
    id: str
    mineDate: str
    articleId: str | None = None
    articleTitle: str
    sourceExcerpt: str = ""
    argumentChain: str = ""
    templateSentence: str = ""
    terms: list[ShenlunMineTermItem] = []
    quotes: list[ShenlunQuoteItem] = []
    verbs: list[ShenlunVerbItem] = []
    argument: ShenlunArgumentSkeleton = ShenlunArgumentSkeleton()
    templates: list[ShenlunTemplateItem] = []
    createdAt: datetime
    updatedAt: datetime


class ShenlunMineLogUpsert(BaseModel):
    mineDate: str | None = None
    articleId: str | None = None
    articleTitle: str = ""
    sourceExcerpt: str = ""
    argumentChain: str = ""
    templateSentence: str = ""
    terms: list[ShenlunMineTermItem | str] = []
    quotes: list[ShenlunQuoteItem] = []
    verbs: list[ShenlunVerbItem] = []
    argument: ShenlunArgumentSkeleton | None = None
    templates: list[ShenlunTemplateItem] = []


class ShenlunMineLogUpdate(BaseModel):
    articleId: str | None = None
    articleTitle: str | None = None
    sourceExcerpt: str | None = None
    argumentChain: str | None = None
    templateSentence: str | None = None
    terms: list[ShenlunMineTermItem | str] | None = None
    quotes: list[ShenlunQuoteItem] | None = None
    verbs: list[ShenlunVerbItem] | None = None
    argument: ShenlunArgumentSkeleton | None = None
    templates: list[ShenlunTemplateItem] | None = None


class ShenlunSkeletonFieldDef(BaseModel):
    key: str
    label: str
    placeholder: str = ""


class ShenlunSkeletonStructure(BaseModel):
    mode: str = "linear"  # linear | points
    fields: list[ShenlunSkeletonFieldDef] = []
    overviewLabel: str = "全文总骨架"
    overviewPlaceholder: str = ""
    pointFields: list[ShenlunSkeletonFieldDef] = []


class ShenlunTermCategoryOut(BaseModel):
    id: str
    name: str
    kind: str = "term"  # term | verb
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunTermCategoryCreate(BaseModel):
    name: str
    kind: str = "term"
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunTermCategoryUpdate(BaseModel):
    name: str | None = None
    kind: str | None = None
    sortOrder: int | None = None
    isEnabled: bool | None = None


class ShenlunSkeletonTemplateOut(BaseModel):
    id: str
    name: str
    description: str = ""
    mode: str = "linear"
    structure: ShenlunSkeletonStructure
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunSkeletonTemplateCreate(BaseModel):
    name: str
    description: str = ""
    mode: str = "linear"
    structure: ShenlunSkeletonStructure | None = None
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunSkeletonTemplateUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    mode: str | None = None
    structure: ShenlunSkeletonStructure | None = None
    sortOrder: int | None = None
    isEnabled: bool | None = None


class ShenlunSentenceTypeOut(BaseModel):
    id: str
    code: str
    name: str
    tip: str = ""
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunSentenceTypeCreate(BaseModel):
    code: str
    name: str
    tip: str = ""
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunSentenceTypeUpdate(BaseModel):
    code: str | None = None
    name: str | None = None
    tip: str | None = None
    sortOrder: int | None = None
    isEnabled: bool | None = None


class ShenlunArgumentMethodOut(BaseModel):
    id: str
    name: str
    scope: str = "point"  # overview | point
    note: str = ""
    template: str = ""
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunArgumentMethodCreate(BaseModel):
    name: str
    scope: str = "point"
    note: str = ""
    template: str = ""
    sortOrder: int = 0
    isEnabled: bool = True


class ShenlunArgumentMethodUpdate(BaseModel):
    name: str | None = None
    scope: str | None = None
    note: str | None = None
    template: str | None = None
    sortOrder: int | None = None
    isEnabled: bool | None = None


class ShenlunMetaOut(BaseModel):
    termCategories: list[ShenlunTermCategoryOut] = []
    verbCategories: list[ShenlunTermCategoryOut] = []
    skeletonTemplates: list[ShenlunSkeletonTemplateOut] = []
    sentenceTypes: list[ShenlunSentenceTypeOut] = []
    argumentMethodPresets: list[ShenlunArgumentMethodOut] = []


class ShenlunNormTermOut(BaseModel):
    id: str
    term: str
    category: str = "其他"
    usageNote: str
    sourceTitle: str
    exampleSentence: str
    articleId: str | None = None
    familiarity: int
    mastered: bool
    createdAt: datetime


class ShenlunNormTermAdd(BaseModel):
    term: str
    category: str = "其他"
    usageNote: str = ""
    sourceTitle: str = ""
    exampleSentence: str = ""
    articleId: str | None = None


class ShenlunNormTermUpdate(BaseModel):
    category: str | None = None
    usageNote: str | None = None
    exampleSentence: str | None = None
    familiarity: int | None = None
    mastered: bool | None = None
    sourceTitle: str | None = None


class ShenlunStatsOut(BaseModel):
    weekMineDays: int
    weekMineTarget: int = 7
    termCount: int
    learningTermCount: int
    todayMined: bool
    weekDrillCount: int = 0


class ShenlunDrillLogOut(BaseModel):
    id: str
    drillType: str
    content: str
    prompt: str
    refMineId: str | None = None
    refTermIds: list[str] = []
    createdAt: datetime


class ShenlunDrillCreate(BaseModel):
    drillType: str  # sentence | imitate | oral
    content: str
    prompt: str = ""
    refMineId: str | None = None
    refTermIds: list[str] = []


RMRB_THEME_TAG_PRESETS = [
    "政绩观",
    "社会治理",
    "乡村振兴",
    "县域经济",
    "高质量发展",
    "民生保障",
    "作风建设",
    "基层减负",
    "科技创新",
    "文化建设",
    "生态文明",
    "依法治国",
]

